File: backend/app/models/project.py
import enum
from datetime import date, datetime
from typing import TYPE_CHECKING, Optional
from sqlalchemy import (
    CheckConstraint,
    Date,
    DateTime,
    Enum,
    ForeignKey,
    Integer,
    String,
    Text,
    func,
)
from sqlalchemy.orm import Mapped, mapped_column, relationship
from app.db.base import Base

if TYPE_CHECKING:
    from app.models.intern import Intern


class ProjectStatus(str, enum.Enum):
    """Enumeration of valid project workflow statuses."""

    NOT_STARTED = "NOT_STARTED"
    IN_PROGRESS = "IN_PROGRESS"
    COMPLETED = "COMPLETED"
    ON_HOLD = "ON_HOLD"


class Project(Base):
    """Project entity model representing a project assigned to an intern."""

    __tablename__ = "projects"
    __table_args__ = (
        CheckConstraint(
            "progress >= 0 AND progress <= 100",
            name="check_project_progress_range",
        ),
    )

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    name: Mapped[str] = mapped_column(String(150), nullable=False)
    description: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    intern_id: Mapped[int] = mapped_column(
        Integer,
        ForeignKey("interns.id", ondelete="RESTRICT"),
        nullable=False,
        index=True,
    )
    start_date: Mapped[date] = mapped_column(Date, nullable=False)
    deadline: Mapped[date] = mapped_column(Date, nullable=False)
    status: Mapped[ProjectStatus] = mapped_column(
        Enum(ProjectStatus, name="project_status_enum", native_enum=False, length=20),
        default=ProjectStatus.NOT_STARTED,
        nullable=False,
    )
    progress: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        nullable=False,
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )

    # Relationships
    intern: Mapped["Intern"] = relationship("Intern", back_populates="projects")

    def __repr__(self) -> str:
        return f"<Project(id={self.id}, name='{self.name}', intern_id={self.intern_id}, status='{self.status}', progress={self.progress})>"

File: frontend/src/types/project.ts
export type ProjectStatus = 'NOT_STARTED' | 'IN_PROGRESS' | 'COMPLETED' | 'ON_HOLD';

export interface Project {
  id: number;
  name: string;
  description: string | null;
  intern_id: number;
  start_date: string;
  deadline: string;
  status: ProjectStatus;
  progress: number;
  created_at?: string;
  updated_at?: string;
}

export interface ProjectCreate {
  name: string;
  description?: string | null;
  intern_id: number;
  start_date: string;
  deadline: string;
  status?: ProjectStatus;
  progress?: number;
}

export interface ProjectUpdate {
  name?: string;
  description?: string | null;
  intern_id?: number;
  start_date?: string;
  deadline?: string;
  status?: ProjectStatus;
  progress?: number;
}

export interface ProjectQueryParams {
  search?: string;
  intern_id?: number | string;
  status?: string;
}
